addons/mail/models/mail_render_mixin.py

```diff
@@ -173,7 +173,7 @@ def _sub_relative2absolute(match):
             # compute here to do it only if really necessary + cache will ensure it is done only once
             # if not base_url
             if not _sub_relative2absolute.base_url:
-                _sub_relative2absolute.base_url = self.env["ir.config_parameter"].sudo().get_param("web.base.url")
+                _sub_relative2absolute.base_url = self.get_base_url()
```

Comment:
In the current code base, that doesn't seem to make sense, or I am missing something.
Indeed, `get_base_url` will always short-circuit and return the ir.config.parameter when `self` is an empty recordset:
```py
    if not self:
        return super().get_base_url()
```
And this method is always called like this `self.env['mail.render.mixin']._replace_local_links(body)`

Reply:
It is also called under `_render_template_postprocess` when `_render_template` is used with `post_process=True`. This usually is called through `_render_field` where the record referenced is not false.

Reply:
Although, now that I'm looking at it, doesn't it make more sense for it to match the `_render_template` pipeline and have the `self.env['mail.render.mixin']._replace_local_links(body)` be replaced with `template._replace_local_links(body)`?

Reply:
> It is also called under _render_template_postprocess

`_render_template_postprocess` is an `api.model` method, `self` should thus be irrelevant.
I don't see a valid reason to do the current change of the PR.

> doesn't it make more sense for it to match the _render_template pipeline and have the self.env['mail.render.mixin']._replace_local_links(body) be replaced with template._replace_local_links(body)?

I don't know, it's not really part of the scope I usually handle, but same rationale as the initial fix of your PR: if you find a bug and can reproduce it, someone can have a look.

Reply:
@rdeodoo 
> `_render_template_postprocess` is an `api.model` method, `self` should thus be irrelevant. I don't see a valid reason to do the current change of the PR.

I looked through the code a bit and I think the confusion comes from the fact that there's a chain of methods that gets called that starts with a method on a recordset.

Whenever you call `send_mail` on a `mail.template`, it calls `generate_email` where it runs `ensure_one()` and calls `_render_field` (one of them being `'body_html'` with `post_process=True`), which calls `_render_template`, which calls `_render_template_postprocess`, which calls `_replace_local_links`

As for the bug:
Install the website_slides and set course to only be valid for one website. If you complete a course, it will send you to the wrong list of courses.

Reply:
Ok I see, so the method is sometime called with a record as `self`, which conveniently could be used to get the base url. While it could make sense (best effort / low cost to try to detect / fix something), it would still be theoretically wrong as this method is not supposed to rely on what's inside `self`.

IMO, it should be the template which should do something like `t-attf-href="{{ object.get_base_url() }}/slides/all"` instead of `href="/slides/all"`, instead of relying of some "magic/lucky" fallback.

Checking quickly a few location in `website_*` modules, there is multiple ways used:
- the one I just suggested, concatenating base url (retrieved on the model) with relative url
  ```xml
  <!-- in website_sale -->
   t-attf-href="{{ object.get_base_url() }}/shop/cart?access_token={{ object.access_token }}"
   ```
- providing a specific base_url in the mail variables, to concat url too
- providing a specific base_url through the context of the record before calling mail method (like `send_mail`)
   ```py
   # in website_slides
   template = record.channel_id.share_template_id.with_context(
        user=self.env.user,
        email=email,
        base_url=record.get_base_url(),  # <- HERE
        fullscreen=fullscreen
   )
   ```
- probably others...

I'd like to summon @tde-banana-odoo which might have some valuable info here, especially since there is some relative URL here and there in the mail templates and I guess they work correctly. There is surely stuff I am missing.

Reply:
While that is technically the correct way to do so, my concern is with the end users who write their own email templates or download themes/email templates from third party stores where the local URL pathing is just assumed to work. That was one of the motivations for me to make this PR in the first place.